Lab1/source/multiphasesorting.cpp

```diff
@@ -0,0 +1,282 @@
+        ip[i] = 1;
+        ms[i] = 1;
+    }
+    ip[countFile - 1] = 0;
+    ms[countFile - 1] = 0;
+}
+
+void recalculate_ip_ms(std::vector<int>& ip, std::vector<int>& ms, int& L, int countFile) {
+    L++; 
+    int ip0 = ip[0];
+    for (int i = 0; i < countFile - 1; ++i) {
+        ms[i] = ip[i + 1] - ip[i] + ip0;
+        ip[i] = ip[i + 1] + ip0;
+    }
+    ms[countFile - 1] = 0; 
+    ip[countFile - 1] = 0; 
+}
+
+void fileSplitting(const std::string &mainFile, int countFile, std::vector<int>& ms, std::vector<int>& ip, std::vector<std::string>& fileNames, int& L) {
+    std::ifstream inputFile(mainFile);
+    if (!inputFile.is_open()) {
+        cerr << "error" << endl;
+        return;
+    }
+    std::vector<std::ofstream> outputFile(countFile);
+    
+    for (int i = 0; i < countFile; ++i) {
+        std::string fileName = nameFile(i);
+        fileNames.push_back(fileName);
+        outputFile[i].open(fileName);
+        if (!outputFile[i].is_open()) {
+            cerr << "error" << endl;
+            return;
+        }
+    }
+    
+    int i = 0;
+    int prev = INT_MIN;
+    int current;
+
+    while (inputFile >> current) {
+        if (current < prev) {
+            outputFile[i] << "\n";
+            ms[i]--;
+            if (i + 1 < ms.size() && ms[i] < ms[i + 1]) {
+                i++;
+            } else if (ms[i] == 0) {
+                recalculate_ip_ms(ip, ms, L, countFile);
+                i = 0;
+            } else {
+                i = 0;
+            }
+        }
+        
+        if (prev != INT_MIN && current >= prev) { 
+            outputFile[i] << " ";
+        }
+           
+        outputFile[i] << current;
+        prev = current;
+    }
+
+    outputFile[i] << "\n";
+    ms[i]--;
+    
+    for (int i = 0; i < countFile; ++i) {
+        outputFile[i].close();
+    }
+
+}
+
+
+bool mergeFiles(int countFile, std::vector<std::string>& fileNames, std::vector<int>& ms, std::vector<int>& ip, int& L) {
+    std::vector<std::fstream*> mainFiles(countFile);
+
+    for(int i = 0; i < countFile - 1; i++){
+        mainFiles[i] = new std::fstream(fileNames[i], std::fstream::in);
+    }
+
+    mainFiles[countFile-1] = new std::fstream(fileNames[countFile - 1], std::fstream::out);
+
+    for(int i = 0; i < countFile; i++){
+        if(!mainFiles[i]->is_open()){
+            return false;
+        }
+    }
+
+    while(L > 0){
+        int minMs = ip[0];
+        for(int i = 0; i < countFile-1; i++){
+            if(ms[i] <= minMs){
+                minMs = ms[i];
+            }
+        }
+
+        ms[countFile-1] += minMs;
+        ip[countFile-1] += minMs;
+
+        for(int i = 0; i < countFile-1; i++){
+            ms[i] -= minMs;
+            ip[i] -= minMs;
+        }
+
+        while(ip[countFile - 2] > 0){
+            std::vector<bool> boolSequences(countFile - 1, false);
+            std::vector<int> elemsOfSequences(countFile - 1);
+            int areSequences = false;
```

Comment:
```suggestion
            bool hasSequences = false;
```
Ну либо оставить `int`, но использоовать его как счётчик - сколько именно последовательностей у нас осталось. По идее, этот вариант позволит избавиться от перебора на 219 строке.